contracts/SupplyChain.sol

```diff
@@ -1,45 +1,52 @@
 pragma solidity 0.5.7;
+pragma experimental ABIEncoderV2;
 
-import "./Package.sol";
-import "./ProducersManager.sol";
+import "./DrugItem.sol";
+import "./VendorsManager.sol";
 
 
-contract SupplyChain is ProducersManager {
+contract SupplyChain is VendorsManager {
 
-    mapping(bytes => Package) private packages;
+    mapping(bytes => DrugItem) private items;
 
-    function getPackage(bytes memory _packageId) public view returns (Package) {
-        return packages[_packageId];
+    function getDrugItem(bytes memory _drugItemId) public view returns (DrugItem) {
+        return items[_drugItemId];
     }
 
-    function registerInitialTransfer(bytes memory _packageId, address _to, Package.ParticipantType _participant)
+    function registerInitialHandover(
+        bytes memory _drugItemId,
+        address _to,
+        DrugItem.ParticipantCategory _participantCategory
+    )
         public
-        onlyNewPackage(_packageId)
-        onlyKnownProducer
+        onlyNewDrugItem(_drugItemId)
+        onlyKnownVendor
     {
-        packages[_packageId] = new Package(_packageId, msg.sender, _to, _participant);
+        items[_drugItemId] = new DrugItem(_drugItemId, msg.sender, _to, _participantCategory);
     }
 
-    function registerTransfer(
-        bytes memory _packageId,
+    function registerHandover(
+        bytes memory _drugItemId,
         address _to,
-        Package.ParticipantType _participant,
+        DrugItem.ParticipantCategory _participantCategory,
         int8 _temperature,
-        Package.TransporterType _transporter
+        DrugItem.TransitCategory _transitCategory
     )
         public
-        onlyKnownPackage(_packageId)
+        onlyKnownDrugItem(_drugItemId)
     {
-        packages[_packageId].logTransfer(msg.sender, _to, _participant, _temperature, _transporter);
+        items[_drugItemId].logHandover(_to, _participantCategory);
+        uint when = items[_drugItemId].getLastHandover().when;
```

Comment:
As you make logTransfer and logTranscitConditions in a single transaction, the stuff with synchronizing when time is useless. Every "now" call in a single transaction would return the same time.